src/compiler/crystal/semantic/exhaustiveness_checker.cr

```diff
@@ -0,0 +1,642 @@
+    type_groups.first.map do |type|
+      compute_target(type).tap do |target|
+        target.add_subtargets(type_groups, 1)
+      end
+    end
+  end
+
+  # Retuens an array of all the types inside `type`:
+  # for unions it's all the union types, otherwise it's just that type.
+  private def expand_types(type)
+    if type.is_a?(UnionType)
+      type.union_types.map(&.devirtualize.as(Type))
+    else
+      [type.devirtualize]
+    end
+  end
+
+  private def compute_target(type : Type)
+    ExhaustivenessChecker.compute_target(type)
+  end
+
+  def self.compute_target(type : Type)
+    case type
+    when BoolType
+      BoolTarget.new(type)
+    when EnumType
+      if type.flags?
+        TypeTarget.new(type)
+      else
+        EnumTarget.new(type)
+      end
+    else
+      TypeTarget.new(type)
+    end
+  end
+
+  private def when_pattern(when_cond) : Pattern?
+    case when_cond
+    when Path
+      if !when_cond.syntax_replacement && !when_cond.target_const &&
+         when_cond.type?
+        # In case of a Path that points to a type,
+        # remove that type from the types we must cover
+        TypePattern.new(when_cond.type.devirtualize)
+      elsif (target_const = when_cond.target_const) &&
+            target_const.namespace.is_a?(EnumType)
+        # If we find a constant that doesn't point to a type (so a value),
+        # if it's an enum member, try to remove it from the targets.
+        EnumMemberPattern.new(target_const)
+      else
+        nil
+      end
+    when Call
+      # Check if it's something like `.foo?` to remove that member from the ones
+      # we must cover.
+      # Note: a user could override the meaning of such methods.
+      # In the future it would be wise to mark these as non-redefinable
+      # so this checks are sounds.
+      if when_cond.obj.is_a?(ImplicitObj) &&
+         when_cond.args.empty? && when_cond.named_args.nil? &&
+         !when_cond.block && !when_cond.block_arg && when_cond.name.ends_with?('?')
+        EnumMemberNamePattern.new(when_cond.name.rchop)
+      else
+        nil
+      end
+    when BoolLiteral
+      BoolPattern.new(when_cond.value)
+    when NilLiteral
+      TypePattern.new(@program.nil_type)
+    when Underscore
+      UnderscorePattern.new
+    else
+      nil
+    end
+  end
+
+  # A type pattern is when you do `when Type`
+  record TypePattern, type : Type
+
+  # An enum member pattern is when you do `when Foo::Bar`
+  # and `Bar` is an enum member of `Foo`
+  record EnumMemberPattern, member : Const
+
+  # An enum member pattern is when you do `when .foo?`
+  record EnumMemberNamePattern, name : String
+
+  # A bool pattern is when you do `when true` or `when false`
+  record BoolPattern, value : Bool
+
+  # An underscore pattern is when you do `when {.., _}` (only in tuple literals)
+  record UnderscorePattern
+
+  alias Pattern = TypePattern | EnumMemberPattern | EnumMemberNamePattern | BoolPattern | UnderscorePattern
```

Comment:
Haskell is getting in here :-)

In this case maybe having a base abstract struct for the record was enough. But I'm not complaining about the pattern in general.

I like that alias of unions of structs can allow reusable record definitions. But here there types are not exposed.

Did you use this pattern because the there is no behavior in these types?

Reply:
Yeah, an abstract struct would have worked too, though there's no data or behavior in common between all the patterns.

Another reason is that once we have exhaustiveness check in place, if I add a new type to `Pattern` then the compiler will tell me where I need to handle it. With inheritance this isn't possible because types can be extended anywhere in the code, even in shards and so on... unless we introduced the concept of a sealed hierarchy.

I actually like being able to code some sections of the code like Haskell/Elm, and having exhaustiveness check will make it even easier and more useful.